internal/elf.go

```diff
@@ -50,3 +52,82 @@ func (se *SafeELFFile) Symbols() (syms []elf.Symbol, err error) {
 	syms, err = se.File.Symbols()
 	return
 }
+
+// DynamicSymbols is the safe version of elf.File.DynamicSymbols.
+func (se *SafeELFFile) DynamicSymbols() (syms []elf.Symbol, err error) {
+	defer func() {
+		r := recover()
+		if r == nil {
+			return
+		}
+
+		syms = nil
+		err = fmt.Errorf("reading ELF dynamic symbols panicked: %s", r)
+	}()
+
+	syms, err = se.File.DynamicSymbols()
+	return
+}
+
+// SymbolsCache is an ELF symbols cache.
+type SymbolsCache struct {
+	cache map[string]map[string]elf.Symbol
+	mu    sync.Mutex
+}
+
+func NewSymbolsCache() *SymbolsCache {
+	var sc SymbolsCache
+	sc.cache = make(map[string]map[string]elf.Symbol)
+	return &sc
+}
+
+func (sc *SymbolsCache) Get(path, symbol string) (*elf.Symbol, error) {
+	symsCache, ok := sc.cache[path]
+	if !ok {
+		return sc.fill(path, symbol)
+	}
+	sym, ok := symsCache[symbol]
+	if !ok {
+		return sc.fill(path, symbol)
+	}
+	return &sym, nil
+}
+
+func (sc *SymbolsCache) fill(path, symbol string) (*elf.Symbol, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("symbols cache: failed to open file: %w", err)
+	}
+	defer f.Close()
+
+	se, err := NewSafeELFFile(f)
+	if err != nil {
+		return nil, fmt.Errorf("symbols cache: failed to parse ELF file: %w", err)
+	}
+
+	// TODO(matt): discuss whether this is the right way to fetch all symbols
```

Comment:
What other ways are there then? Detached debug info?

Reply:
the comment was mostly related about the cache implementation, but I'll ditch the cache